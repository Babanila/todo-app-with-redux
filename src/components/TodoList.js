import React from 'react'
import { useSelector, useDispatch } from 'react-redux'
import { cx, css } from 'emotion'
import SingleTodo from './SingleTodo'
import { completedTodo, removeTodo } from '../actions'

const ulStyles = css`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
`

function TodoList() {
  const todos = useSelector((state) => state.todoReducer.todos)
  const filteredTodos = useSelector((state) => state.visibilityReducer.filteredTodos)
  const [todoDetails, setTodoDetails] = React.useState(todos)

  React.useEffect(() => {
    filteredTodos.length !== 0 ? setTodoDetails(filteredTodos) : setTodoDetails(todos)
  }, [todos, filteredTodos])

  const dispatch = useDispatch()
  const handleClick = (e, id) => dispatch(completedTodo(id))
  const handleDeleteClick = (e, id) => dispatch(removeTodo(id))

  return (
    <ul className={cx(ulStyles)}>
      {todoDetails?.map((items, i) => (
        <SingleTodo key={i} {...items} cancelClick={handleClick} deleteClick={handleDeleteClick} />
      ))}
    </ul>
  )
}

export default TodoList
