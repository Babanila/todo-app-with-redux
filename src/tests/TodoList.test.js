import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import configureStore from 'redux-mock-store'
import TodoList from '../components/TodoList'

const mockStore = configureStore([thunk])

describe('TodoList', () => {
  let store, TodoListWrapper

  beforeEach(() => {
    store = mockStore({
      todoReducer: {
        todos: [
          { id: 1, text: 'Write Code', completed: true },
          { id: 2, text: 'Test Code', completed: false },
          { id: 3, text: 'Refactor Code', completed: true },
          { id: 4, text: 'Review Code', completed: false }
        ]
      },
      visibilityReducer: {
        filteredTodos: [
          { id: 1, text: 'Write Code', completed: true },
          { id: 2, text: 'Test Code', completed: false },
          { id: 3, text: 'Refactor Code', completed: true },
          { id: 4, text: 'Review Code', completed: false }
        ]
      }
    })

    TodoListWrapper = mount(
      <Provider store={store}>
        <TodoList />
      </Provider>
    ).mount()
  })

  it('should be defined', () => {
    expect(TodoList).toBeDefined()
  })

  it('should be a function', () => {
    expect(typeof TodoList).toBe('function')
  })

  it('should renders site title correctly', () => {
    expect(TodoListWrapper).toMatchSnapshot()
  })

  it('should render the exact number of element in component', () => {
    expect(TodoListWrapper.find('ul').length).toEqual(1)
    expect(TodoListWrapper.find('div').length).toEqual(4)
    expect(TodoListWrapper.find('li').length).toEqual(4)
  })
})
