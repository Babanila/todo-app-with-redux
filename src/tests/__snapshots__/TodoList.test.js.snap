// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should renders site title correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <TodoList>
    <ul
      className="css-13stbb8"
    >
      <SingleTodo
        cancelClick={[Function]}
        completed={true}
        deleteClick={[Function]}
        id={1}
        key="0"
        text="Write Code"
      >
        <li
          className="css-l9jawx"
        >
          <div
            onClick={[Function]}
          >
            <input
              checked={true}
              className="css-xpivgo"
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="css-1awclaz"
              style={
                Object {
                  "textDecoration": "line-through",
                }
              }
            >
              Write Code
            </span>
          </div>
          <button
            className="css-1kaf7lk"
            onClick={[Function]}
          >
            Delete
          </button>
        </li>
      </SingleTodo>
      <SingleTodo
        cancelClick={[Function]}
        completed={false}
        deleteClick={[Function]}
        id={2}
        key="1"
        text="Test Code"
      >
        <li
          className="css-l9jawx"
        >
          <div
            onClick={[Function]}
          >
            <input
              checked={false}
              className="css-xpivgo"
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="css-1awclaz"
              style={
                Object {
                  "textDecoration": "none",
                }
              }
            >
              Test Code
            </span>
          </div>
          <button
            className="css-1kaf7lk"
            onClick={[Function]}
          >
            Delete
          </button>
        </li>
      </SingleTodo>
      <SingleTodo
        cancelClick={[Function]}
        completed={true}
        deleteClick={[Function]}
        id={3}
        key="2"
        text="Refactor Code"
      >
        <li
          className="css-l9jawx"
        >
          <div
            onClick={[Function]}
          >
            <input
              checked={true}
              className="css-xpivgo"
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="css-1awclaz"
              style={
                Object {
                  "textDecoration": "line-through",
                }
              }
            >
              Refactor Code
            </span>
          </div>
          <button
            className="css-1kaf7lk"
            onClick={[Function]}
          >
            Delete
          </button>
        </li>
      </SingleTodo>
      <SingleTodo
        cancelClick={[Function]}
        completed={false}
        deleteClick={[Function]}
        id={4}
        key="3"
        text="Review Code"
      >
        <li
          className="css-l9jawx"
        >
          <div
            onClick={[Function]}
          >
            <input
              checked={false}
              className="css-xpivgo"
              onChange={[Function]}
              type="checkbox"
            />
            <span
              className="css-1awclaz"
              style={
                Object {
                  "textDecoration": "none",
                }
              }
            >
              Review Code
            </span>
          </div>
          <button
            className="css-1kaf7lk"
            onClick={[Function]}
          >
            Delete
          </button>
        </li>
      </SingleTodo>
    </ul>
  </TodoList>
</Provider>
`;
